Add menu item and keyCommand support to Web Extension commands.

xeenon opened the pull request:
#### 2a0077acd04eba49f32538cd85b7682d95d0253b
<pre>
Add menu item and keyCommand support to Web Extension commands.
<a href="https://webkit.org/b/265772">https://webkit.org/b/265772</a>
<a href="https://rdar.apple.com/problem/119111895">rdar://problem/119111895</a>

Reviewed by Brian Weinstein.

Adds a menuItem and keyCommand property to _WKWebExtensionCommand for macOS and iOS.
Also adds performCommandForEvent: and commandForEvent: to _WKWebExtensionContext for macOS.
Adds a private _shortcut to _WKWebExtensionCommand for use by Safari.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.mm:
(-[_WKWebExtensionCommand menuItem]): Added.
(-[_WKWebExtensionCommand keyCommand]): Added.
(-[_WKWebExtensionCommand _shortcut]): Added.
(-[_WKWebExtensionCommand _matchesEvent:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommandPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext performCommandForEvent:]): Added.
(-[_WKWebExtensionContext commandForEvent:]): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(+[_WKWebExtensionKeyCommand commandWithTitle:image:input:modifierFlags:handler:]): Added.
(-[_WKWebExtensionKeyCommand copyWithZone:]):
(-[_WKWebExtensionKeyCommand _resolvedTargetFromFirstTarget:]):
(-[_WKWebExtensionKeyCommand _performWebExtensionKeyCommand:]):
(WebKit::WebExtensionCommand::platformMenuItem const): Added.
(WebKit::WebExtensionCommand::keyCommand const): Added.
(WebKit::WebExtensionCommand::matchesEvent const): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::performCommand): Added.
(WebKit::WebExtensionContext::command): Added.
(WebKit::WebExtensionContext::performMenuItem): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
(-[_WKWebExtensionMenuItem initWithTitle:handler:]):
(-[_WKWebExtensionMenuItem copyWithZone:]): Added.
(-[_WKWebExtensionMenuItem _performAction:]):
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICommands.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/271532@main">https://commits.webkit.org/271532@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab17bf7274850067c0ea91ad17d0ea7b50ddb827

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28710 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/7353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30081 "Failed to checkout and rebase branch from PR 21249") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31329 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/29280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9474 "Failed to checkout and rebase branch from PR 21249") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/4713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28980 "Failed to checkout and rebase branch from PR 21249") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/9474 "Failed to checkout and rebase branch from PR 21249") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/30081 "Failed to checkout and rebase branch from PR 21249") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/5283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/9474 "Failed to checkout and rebase branch from PR 21249") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/30081 "Failed to checkout and rebase branch from PR 21249") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/32667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/9474 "Failed to checkout and rebase branch from PR 21249") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/30081 "Failed to checkout and rebase branch from PR 21249") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/32667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/4713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/32667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7018 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/30081 "Failed to checkout and rebase branch from PR 21249") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3706 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/5919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.h:
```diff
@@ -51,7 +51,7 @@ NS_SWIFT_NAME(WKWebExtension.Command)
 @property (nonatomic, readonly, weak) _WKWebExtensionContext *webExtensionContext;
 
-/*! @abstract Unique identifier for the command. */
+/*! @abstract An unique identifier for the command. */
```
nit: `A unique`

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
```diff
@@ -164,6 +205,90 @@
+
+    return false;
+}
+#endif
```
`#endif // USE(APPKIT)`

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICommands.mm:
```diff
@@ -120,6 +121,140 @@
+{
+    auto *backgroundScript = Util::constructScript(@[
+        @"browser.commands.onCommand.addListener((command, tab) => {",
+        @"  browser.test.assertEq(command, 'test-command', 'The command should be')",
```
`@"  browser.test.assertEq(command, 'test-command', 'The command should be test-command')",`?

kiaraarose replied:
seems like this and the ones below are missing more text for what the equality should be

b-weinstein replied:
I believe the expected result gets put into the test output if it fails.

xeenon replied:
Yeah, i had been being overly verbose with the message, it can be shorter like the new values, if a message is needed at all.

xeenon committed:
Add menu item and keyCommand support to Web Extension commands.
https://webkit.org/b/265772
rdar://problem/119111895

Reviewed by Brian Weinstein.

Adds a menuItem and keyCommand property to _WKWebExtensionCommand for macOS and iOS.
Also adds performCommandForEvent: and commandForEvent: to _WKWebExtensionContext for macOS.
Adds a private _shortcut to _WKWebExtensionCommand for use by Safari.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.mm:
(-[_WKWebExtensionCommand menuItem]): Added.
(-[_WKWebExtensionCommand keyCommand]): Added.
(-[_WKWebExtensionCommand _shortcut]): Added.
(-[_WKWebExtensionCommand _matchesEvent:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommandPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext performCommandForEvent:]): Added.
(-[_WKWebExtensionContext commandForEvent:]): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(+[_WKWebExtensionKeyCommand commandWithTitle:image:input:modifierFlags:handler:]): Added.
(-[_WKWebExtensionKeyCommand copyWithZone:]):
(-[_WKWebExtensionKeyCommand _resolvedTargetFromFirstTarget:]):
(-[_WKWebExtensionKeyCommand _performWebExtensionKeyCommand:]):
(WebKit::WebExtensionCommand::platformMenuItem const): Added.
(WebKit::WebExtensionCommand::keyCommand const): Added.
(WebKit::WebExtensionCommand::matchesEvent const): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::performCommand): Added.
(WebKit::WebExtensionContext::command): Added.
(WebKit::WebExtensionContext::performMenuItem): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
(-[_WKWebExtensionMenuItem initWithTitle:handler:]):
(-[_WKWebExtensionMenuItem copyWithZone:]): Added.
(-[_WKWebExtensionMenuItem _performAction:]):
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICommands.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/271532@main
